### ISSUE TITLE
Property's default sysdate() cause exception on insert

### ISSUE THREAD

**Issue by ivmakk** [orientechnologies/orientdb#6333]
### Expected
- Create record with current date
### Actual:

```
Error: com.orientechnologies.orient.core.exception.ODatabaseException: Error during saving of record with rid #12:-1
    DB name="webtest01"
    DB name="webtest01"

Error: com.orientechnologies.orient.core.exception.ODatabaseException: Error on saving record in cluster #12
    DB name="webtest01"

Error: java.lang.ClassCastException: com.orientechnologies.orient.core.sql.filter.OSQLFilterItemField cannot be cast to java.util.Date
```

Also looks like instead of storing function as `DEFAULT` value for `Users.createdAt` property, it stored result of executing `sysdate()` function:

```
CLASS 'Users'

Records..............: 3
Default cluster......: users (id=12)
Supported clusters...: users(12)
Cluster selection....: round-robin
Oversize.............: 0.0

PROPERTIES
+----+---------+-----------------+---------+--------+--------+----+----+-------+-----------------------------+
|#   |NAME     |LINKED-TYPE/CLASS|MANDATORY|READONLY|NOT-NULL|MIN |MAX |COLLATE|DEFAULT                      |
+----+---------+-----------------+---------+--------+--------+----+----+-------+-----------------------------+
|0   |createdAt|                 |false    |false   |false   |    |    |default|Tue Jun 21 16:50:58 EEST 2016|

```
## Steps to reproduce the problem

```
CREATE CLASS Users
CREATE PROPERTY Users.email STRING
CREATE PROPERTY Users.createdAt DATETIME
ALTER PROPERTY Users.createdAt DEFAULT sysdate()
INSERT INTO Users SET email="mail@mail.com"
```
### Runninng Mode
- Remote
### Misc
- I have a distributed setup with single server.
### OrientDB Version
- 2.2.2
### Operating System
- Linux
### Java Version
- 8

**Comment by tglman** [comment]
hi @ivmakk,

there is a workaround for this today and is using `ALTER PROPERTY Users.createdAt DEFAULT "sysdate()" ` we are in the meanwhile checking this feature, due to the fact that was a communite provided feature and we probably didn't check correctly all the possible case this feature would have impacted. 

Regards

**Comment by tglman** [comment]
hi @ivmakk,

Did you have a change to try it ? can we close this ?

**Comment by xscratt** [comment]
Hi @tglman,

I just checked this in OrientDB 2.2.3, and seems it still reproduce if write `sysdate()` without quotes:
```
orientdb {db=test}> INSERT INTO Users SET email="mail@mail.com"

Error: com.orientechnologies.orient.core.exception.ODatabaseException: Error during saving of record with rid #21:-1
    DB name="test"
    DB name="test"

Error: com.orientechnologies.orient.core.exception.ODatabaseException: Error on saving record in cluster #21
    DB name="test"

Error: java.lang.ClassCastException: com.orientechnologies.orient.core.sql.filter.OSQLFilterItemField cannot be cast to java.util.Date

orientdb {db=test}> info class Users

CLASS 'Users'

Records..............: 0
Default cluster......: users (id=21)
Supported clusters...: users(21), users_1(22), users_2(23), users_3(24)
Cluster selection....: round-robin
Oversize.............: 0.0

PROPERTIES
+----+---------+-----------------+---------+--------+--------+----+----+-------+----------------------------+
|#   |NAME     |LINKED-TYPE/CLASS|MANDATORY|READONLY|NOT-NULL|MIN |MAX |COLLATE|DEFAULT                     |
+----+---------+-----------------+---------+--------+--------+----+----+-------+----------------------------+
|0   |createdAt|                 |false    |false   |false   |    |    |default|Sat Jul 02 12:51:43 GMT 2016|
|1   |email    |                 |false    |false   |false   |    |    |default|                            |
+----+---------+-----------------+---------+--------+--------+----+----+-------+----------------------------+
```
The proposed workaround with quotes (`"sysdate()"`) is working, but in OrientDB 2.2.0 both cases works as expected (whether or not use quotes in `ALTER PROPERTY Users.createdAt DEFAULT sysdate()` command) -- the schema contains `sysdate()` as a default value of the property (instead of the function's execution result):
```
orientdb {db=test}> info class Users

CLASS 'Users'

Records..............: 1
Default cluster......: users (id=21)
Supported clusters...: users(21), users_1(22), users_2(23), users_3(24)
Cluster selection....: round-robin
Oversize.............: 0.0

PROPERTIES
+----+---------+-----------------+---------+--------+--------+----+----+-------+---------+
|#   |NAME     |LINKED-TYPE/CLASS|MANDATORY|READONLY|NOT-NULL|MIN |MAX |COLLATE|DEFAULT  |
+----+---------+-----------------+---------+--------+--------+----+----+-------+---------+
|0   |createdAt|                 |false    |false   |false   |    |    |default|sysdate()|
|1   |email    |                 |false    |false   |false   |    |    |default|         |
+----+---------+-----------------+---------+--------+--------+----+----+-------+---------+
```
So as for me it seems one of these versions contains a bug (2.2.0 or 2.2.2+).

**Comment by luigidellaquila** [comment]
Hi @xscratt 

We have a problem here, and I think there is no solution in the short term.



The DEFAULT value could be evaluated in three different moments:
- 1. at ALTER PROPERTY statement execution time. In this case the value of sysdate() is calculated only once and stored in the schema, so every time you create a new record, you will receive as a default value the formerly calculated value (always the same). This is what happens now when you do 
`ALTER PROPERTY Users.createdAt DEFAULT sysdate()` without string quotes
- 2. at INSERT statement execution time. In this case the function is evaluated every time you do an INSERT INTO statement. This is what happens when you do `ALTER PROPERTY Users.createdAt DEFAULT "sysdate()"` with string quotes
- 3. never. eg. if it was supported, a `ALTER PROPERTY Users.createdAt DEFAULT "sysdate()"` would produce the exact "sysdate()" string as a default value for properties. Now it's just not supported (unfortunately)

Technically speaking, here we have two low level problems:
- the DEFAULT attribute in the schema is a STRING, so sysdate() can be stored only as a string (all the "custom" values are strings), so I cannot distinguish between the "sysdate()" string and the "sysdate()" function to be evaluated at INSERT time (no way to implement both 2 and 3). Doing it simple, in case n. 2 I just store the string "sysdate()" in the schema, but in case n. 1 I have no alternative: I can pre-calculate the value and store it (that is what happens now) or I can store the "sysdate()" string. This second case would be exactly the same as n 2, so it does not make much sense.
- we have a single attribute (`defaut`) for two behaviors, in reality we would need two, eg. `default_value`, that is just used as is, and `default_expression` that is evaluated at INSERT time 

To make an example of what I mean :

```
(not supported yet)

CREATE PROPERTY Foo.bar STRING DEFAULT_VALUE "a + b"
```

So in short:
- is current behavior correct and complete? No, of course 
- is there a way to fix this behavior to support all the three cases now? Unfortunately the answer is no (see above)
- is there a plan to change it? Yes, but I'm afraid it won't be in the very short term.


-

**Comment by xscratt** [comment]
Hi, @luigidellaquila 

Thank you for the detailed response - I was thinking about this and agree that implementation with only one `DEFAULT` property is ambiguous and cannot satisfy all possible cases and expectations. So probably the best solution now - really leave current implementation 'as is' for a while.

But anyway, in my opinion the reason of this exception - the result of `sysdate()`, computed at `ALTER PROPERTY` execution time, doesn't use the database's datetime format (`Sat Jul 02 12:51:43 GMT 2016` does not match `yyyy-MM-dd HH:mm:ss`), and maybe it makes sense to fix only this, just to make current behavior more consistent.

Thank you.

**Comment by ivmakk** [comment]
@tglman, thank you, it works, but I only need to know, that this behaver (passing quoted function as DEFAULT value) will not have been broken in next versions.
@luigidellaquila great problem description, thanx.

As possible solution, why not to add constant(s) as default value, like `DEFAULT CURRENT_TIMESTAMP` or `DEFAULT SYSDATE`? Is it to hard to add that to parsing rules?

**Comment by luigidellaquila** [comment]
Hi guys,

I'll see if I can do something to make it a bit easier...
Probably there will be some changes in v3.0 about the API (eg. two different keywords, like DEFAULT_VALUE and DEFAULT_FUNCTION), but it's not defined at the moment

Thanks

Luigi

**Comment by a-unite** [comment]
Possibly it will be enough to add DEF_FUNCTION keyword in addition to already well established DEFAULT?
